PG-1004: Skip encrypted tables in `pg_checksums`

jeltz opened the pull request:
I want to check if I am going in the right direction with this.

This PR makes `pg_checksums` skip encrypted relations by looking if the relation has a SMGR key. The patch is a bit ugly but I do not think it can be cleaned up that much outside coming up with better function names.

We print log messages about skipping the tables only when `--verbose` is passed. Maybe that is bad, maybe we should either provide a summary of how many `pg_tde` relations we found. Because right now it is quite hidden.

```
$ bin/pg_checksums data --verbose
pg_checksums: checksums verified in file "data/global/1213_fsm"
pg_checksums: checksums verified in file "data/global/1261_vm"
pg_checksums: checksums verified in file "data/global/2964"
pg_checksums: checksums verified in file "data/global/4181"
[... 548 lines ...]
pg_checksums: checksums verified in file "data/base/5/3079"
pg_checksums: checksums verified in file "data/base/5/3502"
pg_checksums: skipped pg_tde encrypted file "data/base/5/16479"
pg_checksums: checksums verified in file "data/base/5/4154"
pg_checksums: checksums verified in file "data/base/5/2687"
pg_checksums: checksums verified in file "data/base/5/2600_fsm"
pg_checksums: skipped pg_tde encrypted file "data/base/5/16457"
pg_checksums: checksums verified in file "data/base/5/3603_fsm"
pg_checksums: checksums verified in file "data/base/5/2615_fsm"
[... 395 lines ...]
pg_checksums: checksums verified in file "data/base/1/2654"
pg_checksums: checksums verified in file "data/base/1/2673"
pg_checksums: checksums verified in file "data/base/1/2610_vm"
pg_checksums: checksums verified in file "data/base/1/2602_vm"
pg_checksums: skipped pg_tde encrypted file "data/pg_tblspc/16466/PG_17_202406281/5/16474"
pg_checksums: skipped pg_tde encrypted file "data/pg_tblspc/16466/PG_17_202406281/5/16477"
pg_checksums: skipped pg_tde encrypted file "data/pg_tblspc/16466/PG_17_202406281/5/16478"
Checksum operation completed
Files scanned:   948
Blocks scanned:  2858
Bad checksums:  0
Data checksum version: 1
```

TODO:

- ~~Pick better function names~~
- ~~Fix autotools support~~
- Write tests?
- Improve output?

jeltz committed:
PG-1004 Remove unnecessary includes in pg_tde_tdemap.h

These includes were not used and made the file impossible to include in
frontend code.

jeltz committed:
PG-1004 Remove pg_tde from pg_checksums exclude list

The exclude list is for files only, there is no need to add a directory
to the list.

jeltz committed:
PG-1004 Skip encrypted files in pg_checksums

We check if a file is encrypted by looking at if it has a storage
manager encryption key for the relation file and if it has a key
we skip it.

These lookups require use to retrieve the tablespace and database
oids from the path to to the relation file.

dAdAbird commented on src/bin/pg_checksums/pg_checksums.c:
```diff
@@ -592,14 +641,14 @@ main(int argc, char *argv[])
-		(void) scan_directory(DataDir, "pg_tblspc", false);
+		(void) scan_directory(DataDir, "global", GLOBALTABLESPACE_OID, false);
+		(void) scan_directory(DataDir, "base", DEFAULTTABLESPACE_OID, false);
+		(void) scan_directory(DataDir, "pg_tblspc", InvalidOid, false);
```
There is `Assert(tablespace != InvalidOid);` in scan_directory

jeltz replied:
Yes, that is exactly what the assert is supposed to guard against. The code should always have changed the value of the parameter before we reach the assert.

dutow replied:
> The code should always have changed the value of the parameter before we reach the assert.

Where? I also don't see this, the `Assert` is the first reference of the `tablespace` parameter in `scan_directory`.

jeltz replied:
When recursing into a directory under `pg_tblspc`.

```
			if (strncmp("pg_tblspc", subdir, strlen("pg_tblspc")) == 0)
			{
				char		tblspc_path[MAXPGPATH];
				struct stat tblspc_st;

				/*
				 * Resolve tablespace location path and check whether
				 * TABLESPACE_VERSION_DIRECTORY exists.  Not finding a valid
				 * location is unexpected, since there should be no orphaned
				 * links and no links pointing to something else than a
				 * directory.
				 */
				snprintf(tblspc_path, sizeof(tblspc_path), "%s/%s/%s",
						 path, de->d_name, TABLESPACE_VERSION_DIRECTORY);

				if (lstat(tblspc_path, &tblspc_st) < 0)
					pg_fatal("could not stat file \"%s\": %m",
							 tblspc_path);

				/*
				 * Move backwards once as the scan needs to happen for the
				 * contents of TABLESPACE_VERSION_DIRECTORY.
				 */
				snprintf(tblspc_path, sizeof(tblspc_path), "%s/%s",
						 path, de->d_name);

				/* Looks like a valid tablespace location */
				dirsize += scan_directory(tblspc_path,
										  TABLESPACE_VERSION_DIRECTORY,
										  atooid(de->d_name),
										  sizeonly);
			}
```

dAdAbird reviewed (commented):
The idea overall LGTM. I couldn't think of better way to approach the issue.

BTW, does it check WAL segments as well?

jeltz commented:
@dAdAbird `pg_checksums` only looks at data files and does not look at the WAL.

jeltz committed:
PG-1004 Refactor initialization of TDE in frontend code

- Do not assume that people always will need xlog and smgr
- Use an inline function instead of a macro
- Make the initialization of the path more explicit

jeltz committed:
PG-1004 Always use define when referring to pg_tde path

jeltz commented:
Cleaned it up a bit so now it should be ready for review. I still wonder if the output needs to be improved.
